src/lib_c/x86_64-windows-msvc/c/iocp.cr

```diff
@@ -0,0 +1,53 @@
+require "c/winnt"
+
+# WSAOVERLAPPED is the primary communication structure for async I/O on Windows.
+# See https://docs.microsoft.com/en-us/windows/win32/api/winsock2/ns-winsock2-wsaoverlapped
+@[Extern]
+struct WSAOVERLAPPED
+  internal : LibC::ULONG_PTR
+  internalHigh : LibC::ULONG_PTR
+  offset : LibC::DWORD
+  offsetHigh : LibC::DWORD
+  hEvent : LibC::HANDLE
+  property cEvent : Void*
+
+  def initialize(crystal_event : Crystal::Event)
+    @cEvent = crystal_event.unsafe_as(Pointer(Void))
```

Comment:
This will work only when the size of `Crystal::Event` is less than or equal to 8.

`Crystal::Event` is converted to `Pointer(Void)` here and reverted by the following code in `event_loop_iocp.cr`
```crystal
       next_event = io_entry.first.lpOverlapped.value.cEvent.unsafe_as(Crystal::Event)
```
I suspected that it reverted only the first 8 bytes. So I wrote the following code and ran it.
```crystal
@[Extern]
struct DummyEvent
  property buf : UInt8[32] = StaticArray(UInt8, 32).new { |i| i.to_u8 }
end

crystal_event = DummyEvent.new
puts(crystal_event)
cEvent = crystal_event.unsafe_as(Pointer(Void))
puts(cEvent.unsafe_as(DummyEvent))
```
This printed:
```
DummyEvent(@buf=StaticArray[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30
, 31])
DummyEvent(@buf=StaticArray[0, 1, 2, 3, 4, 5, 6, 7, 7, 225, 85, 153, 227, 1, 0, 0, 10, 254, 117, 233, 45, 0, 0, 0, 9, 119, 86, 153, 227, 1, 
0, 0])
```
Only the first 8 bytes were reverted. The rest bytes seem uninitialized ones.

Reply:
It's been a long time since I worked with pointers in Crystal, but it seems like this gives the correct result:

Store the pointer to `Crystal::Event`:

```crystal
def initialize(crystal_event : Crystal::Event)
  @cEvent = pointerof(crystal_event)
  ...
```

Get the the pointed-to value back:

```crystal
next_event = io_entry.first.lpOverlapped.value.cEvent.value
```

Reply:
When you do `unsafe_as(Pointer(Void))` the value is interpreted as a pointer. That doesn't work because `crystal_event` is not a pointer, but the actual value.

`pointerof(crystal_event)` isn't reliable because it points at the stack which will be invalidated once the `initialize` method returns. You either need to ensure that the stack pointer is valid during the lifetime of its use (not sure if that's the case here), or put it on the heap.